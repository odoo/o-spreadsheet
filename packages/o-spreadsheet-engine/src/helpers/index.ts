export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
