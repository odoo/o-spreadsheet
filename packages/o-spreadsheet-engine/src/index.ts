export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
