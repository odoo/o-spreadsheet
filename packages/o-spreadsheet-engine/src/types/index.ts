export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
