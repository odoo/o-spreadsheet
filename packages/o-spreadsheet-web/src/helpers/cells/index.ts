export * from "../links";
export * from "./cell_evaluation";
