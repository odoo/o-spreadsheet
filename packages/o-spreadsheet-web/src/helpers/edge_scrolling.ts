export const MAX_DELAY = 140;
export const MIN_DELAY = 20;
const ACCELERATION = 0.035;

/**
 * Decreasing exponential function used to determine the "speed" of edge-scrolling
 * as the timeout delay.
 *
 * Returns a timeout delay in milliseconds.
 */
export function scrollDelay(value: number): number {
  // decreasing exponential from MAX_DELAY to MIN_DELAY
  return MIN_DELAY + (MAX_DELAY - MIN_DELAY) * Math.exp(-ACCELERATION * (value - 1));
}
