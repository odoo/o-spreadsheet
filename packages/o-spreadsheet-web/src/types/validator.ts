import { CommandResult } from "./commands";
import { Validation } from "./misc";

export interface Validator {
  /**
   * Combine multiple validation functions into a single function
   * returning the list of result of every validation.
   */
  batchValidations<T>(...validations: Validation<T>[]): Validation<T>;

  /**
   * Combine multiple validation functions. Every validation is executed one after
   * the other. As soon as one validation fails, it stops and the cancelled reason
   * is returned.
   */
  chainValidations<T>(...validations: Validation<T>[]): Validation<T>;

  checkValidations<T>(command: T, ...validations: Validation<T>[]): CommandResult | CommandResult[];
}
