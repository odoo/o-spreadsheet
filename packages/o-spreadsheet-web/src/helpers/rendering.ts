import { GridRenderingContext, Highlight, Rect } from "../types";
import { toHex } from "./color";

import { HIGHLIGHT_COLOR } from "../constants";
import { setColorAlpha } from "./color";

export function drawHighlight(
  renderingContext: GridRenderingContext,
  highlight: Highlight,
  rect: Rect
) {
  const { x, y, width, height } = rect;
  if (width < 0 || height < 0) {
    return;
  }
  const color = highlight.color || HIGHLIGHT_COLOR;

  const { ctx } = renderingContext;
  ctx.save();
  if (!highlight.noBorder) {
    if (highlight.dashed) {
      ctx.setLineDash([5, 3]);
    }
    ctx.strokeStyle = color;
    if (highlight.thinLine) {
      ctx.lineWidth = 1;
      ctx.strokeRect(x, y, width, height);
    } else {
      ctx.lineWidth = 2;
      /** + 0.5 offset to have sharp lines. See comment in {@link RendererPlugin#drawBorder} for more details */
      ctx.strokeRect(x + 0.5, y + 0.5, width, height);
    }
  }
  if (!highlight.noFill) {
    ctx.fillStyle = setColorAlpha(toHex(color), highlight.fillAlpha ?? 0.12);
    ctx.fillRect(x, y, width, height);
  }
  ctx.restore();
}
