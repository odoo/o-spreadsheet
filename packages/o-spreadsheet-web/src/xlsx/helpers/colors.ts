import { toHex } from "../../helpers";
import { Color } from "../../types";
import { XlsxHexColor } from "../../types/xlsx";

/**
 * Convert a JS color hexadecimal to an excel compatible color.
 *
 * In Excel the color don't start with a '#' and the format is AARRGGBB instead of RRGGBBAA
 */
export function toXlsxHexColor(color: Color): XlsxHexColor {
  color = toHex(color).replace("#", "");
  // alpha channel goes first
  if (color.length === 8) {
    return color.slice(6) + color.slice(0, 6);
  }
  return color;
}
