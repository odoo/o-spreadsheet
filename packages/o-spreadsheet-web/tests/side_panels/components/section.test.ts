import { Component, xml } from "@odoo/owl";
import { Section } from "../../../src/components/side_panel/components/section/section";
import { SpreadsheetChildEnv } from "../../../src/types";
import { mountComponent } from "../../test_helpers/helpers";

let fixture: HTMLElement;

type Props = Section["props"];

describe("Section", () => {
  test("Can render a section without a title", async () => {
    class SectionContainer extends Component<Props, SpreadsheetChildEnv> {
      static template = xml/* xml */ `
    <div class="container">
      <Section t-props="props">
        <div class="content">My content</div>
      </Section>
    </div>
  `;
      static components = { Section };
      static props = { class: String };
    }
    const props = { class: "my-class" };
    ({ fixture } = await mountComponent(SectionContainer, { props }));
    expect(fixture).toMatchSnapshot();
  });

  test("Can render a section with a title slot", async () => {
    class SectionContainer extends Component<Props, SpreadsheetChildEnv> {
      static template = xml/* xml */ `
    <div class="container">
      <Section t-props="props">
        <t t-set-slot="title">My title</t>
        <div class="content">My content</div>
      </Section>
    </div>
  `;
      static components = { Section };
      static props = { class: String };
    }
    const props = { class: "my-class" };
    ({ fixture } = await mountComponent(SectionContainer, { props }));
    expect(fixture).toMatchSnapshot();
  });

  test("Can render a section with a title props", async () => {
    class SectionContainer extends Component<Props, SpreadsheetChildEnv> {
      static template = xml/* xml */ `
    <div class="container">
      <Section t-props="props" title.translate="My title">
        <div class="content">My content</div>
      </Section>
    </div>
  `;
      static components = { Section };
      static props = { class: String };
    }
    const props = { class: "my-class" };
    ({ fixture } = await mountComponent(SectionContainer, { props }));
    expect(fixture).toMatchSnapshot();
  });

  test("Can render a section with both title props and slot", async () => {
    class SectionContainer extends Component<Props, SpreadsheetChildEnv> {
      static template = xml/* xml */ `
    <div class="container">
      <Section t-props="props" title.translate="My title from props">
      <t t-set-slot="title">My title from slot</t>
        <div class="content">My content</div>
      </Section>
    </div>
  `;
      static components = { Section };
      static props = { class: String };
    }
    const props = { class: "my-class" };
    ({ fixture } = await mountComponent(SectionContainer, { props }));
    expect(fixture).toMatchSnapshot();
  });
});
