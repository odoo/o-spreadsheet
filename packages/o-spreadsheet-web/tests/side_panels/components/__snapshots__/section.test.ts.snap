// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section Can render a section with a title props 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="o-section my-class"
    >
      <div
        class="o-section-title"
      >
        My title
        
        
      </div>
      
      <div
        class="content"
      >
        My content
      </div>
      
    </div>
  </div>
</div>
`;

exports[`Section Can render a section with a title slot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="o-section my-class"
    >
      <div
        class="o-section-title"
      >
        
        My title
        
      </div>
      
      <div
        class="content"
      >
        My content
      </div>
      
    </div>
  </div>
</div>
`;

exports[`Section Can render a section with both title props and slot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="o-section my-class"
    >
      <div
        class="o-section-title"
      >
        My title from props
        My title from slot
        
      </div>
      
      <div
        class="content"
      >
        My content
      </div>
      
    </div>
  </div>
</div>
`;

exports[`Section Can render a section without a title 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="o-section my-class"
    >
      
      <div
        class="content"
      >
        My content
      </div>
      
    </div>
  </div>
</div>
`;
